src/celeritas/global/ActionLauncher.device.hh

```diff
@@ -62,25 +106,39 @@ launch_action_impl(Range<ThreadId> const thread_range, F execute_thread)
  }
  * \endcode
  */
-template<class F>
+template<class F, int... bounds>
```

Comment:
Is there a way to make the additional arguments default to properties of the `F` functor if they're available? I.e. if `F` defines a `inline static constexpr int max_blocks = ...;` then use it, and if it has `min_warps_per_eu` then use that as well?

And should we default the max block size to `Device::max_block_size` if unspecified, or leave it to the compiler?

Reply:
Mmh maybe we can use SFINAE  to have an implementation for cases where F::max_blocks exists, and cases where it doesn't?

I don't think we should have a default for max_block_size because CUDA (and probably HIP as well) will constrain the registers per kernel as follow: 

> the compiler first derives from them the upper limit L on the number of registers the kernel should use to ensure that minBlocksPerMultiprocessor blocks (or a single block if minBlocksPerMultiprocessor is not specified) of maxThreadsPerBlock threads can reside on the multiprocessor

so specifying a default value has drawbacks; if we specify 1024 which is the max block size then we constraint the register usage, and 256 prevents us from using a larger blocks

Reply:
We don't use larger blocks at runtime, so we should probably set these defaults on a case-by-case basis.

Reply:
I got an implementation  working where the ActionLauncher automatically picks up on the `static constexpr` though it's a lot of meta-programming and quite verbose.